op-node/node/builder.go

```diff
@@ -0,0 +1,157 @@
+package node
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"time"
+
+	builderSpec "github.com/attestantio/go-builder-client/spec"
+	consensusspec "github.com/attestantio/go-eth2-client/spec"
+
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/pkg/errors"
+)
+
+var (
+	errHTTPErrorResponse = errors.New("HTTP error response")
+)
+
+const PathGetPayload = "/eth/v1/builder/payload"
+
+type BuilderAPIConfig struct {
+	Timeout  time.Duration
+	Endpoint string
+}
+
+type BuilderAPIClient struct {
+	log        log.Logger
+	config     *BuilderAPIConfig
+	httpClient *client.BasicHTTPClient
+}
+
+func NewBuilderClient(log log.Logger, endpoint string, timeout time.Duration) *BuilderAPIClient {
+	httpClient := client.NewBasicHTTPClient(endpoint, log)
+	config := &BuilderAPIConfig{
+		Timeout:  timeout,
+		Endpoint: endpoint,
+	}
+
+	return &BuilderAPIClient{
+		httpClient: httpClient,
+		config:     config,
+		log:        log,
+	}
+}
+
+func (s *BuilderAPIClient) Enabled() bool {
+	return true
+}
+
+func (s *BuilderAPIClient) Timeout() time.Duration {
+	return s.config.Timeout
+}
+
+type httpErrorResp struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+}
+
+func (s *BuilderAPIClient) GetPayload(ctx context.Context, ref eth.L2BlockRef, log log.Logger) (*eth.ExecutionPayloadEnvelope, *big.Int, error) {
+	responsePayload := new(builderSpec.VersionedSubmitBlockRequest)
+	slot := ref.Number + 1
+	parentHash := ref.Hash
+	url := fmt.Sprintf("%s/%d/%s", PathGetPayload, slot, parentHash.String())
+	header := http.Header{"Accept": {"application/json"}}
+	resp, err := s.httpClient.Get(ctx, url, nil, header)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	defer resp.Body.Close()
+
+	bodyBytes, err := io.ReadAll(resp.Body)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, nil, errHTTPErrorResponse
+	}
+
+	if err := json.Unmarshal(bodyBytes, responsePayload); err != nil {
+		return nil, nil, err
+	}
+
+	if responsePayload.Version != consensusspec.DataVersionDeneb {
```

Comment:
is it possible to detect what fork version the response is meant to be based on chain config instead of hardcoding to deneb? This will make it flexible for upcoming hard forks

Reply:
then there shouldn't be a reason to check the versioning? If the attestant types change the versioning would update as well

Reply:
this is to guard against any nil pointer reference error.